File: gin/internal/model/tournaments/tournament_round.go
package model_tournaments

import (
	"time"

	"gorm.io/gorm"
	"fmt"
)

type TournamentRoundStatusType string
const (
	TournamentRoundStatusType_Pending TournamentRoundStatusType = "Pending"
	TournamentRoundStatusType_Active TournamentRoundStatusType = "Active"
	TournamentRoundStatusType_Completed TournamentRoundStatusType = "Completed"
)

// TournamentRoundCreateRequest is the POST body.
type TournamentRoundCreateRequest struct {
	RoundNumber int `json:"round_number"`
	Status TournamentRoundStatusType `json:"status" binding:"required"`
	StartedAt *string `json:"started_at"`
	EndedAt *string `json:"ended_at"`
	TimeLimitMinutes int `json:"time_limit_minutes"`
	TournamentID uint `json:"tournament_id"`
}

// TournamentRoundUpdateRequest is the PUT/PATCH body — all fields optional.
type TournamentRoundUpdateRequest struct {
	RoundNumber *int `json:"round_number"`
	Status *TournamentRoundStatusType `json:"status"`
	StartedAt *string `json:"started_at"`
	EndedAt *string `json:"ended_at"`
	TimeLimitMinutes *int `json:"time_limit_minutes"`
	TournamentID *uint `json:"tournament_id"`
}

// TournamentRoundResponse is the JSON representation returned by the API.
type TournamentRoundResponse struct {
	ID        uint      `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	RoundNumber int `json:"round_number"`
	Status TournamentRoundStatusType `json:"status"`
	StartedAt *string `json:"started_at"`
	EndedAt *string `json:"ended_at"`
	TimeLimitMinutes int `json:"time_limit_minutes"`
	TournamentID uint `json:"tournament_id"`
}

type TournamentRound struct {
	gorm.Model
	RoundNumber int `gorm:"column:round_number;not null"`
	Status TournamentRoundStatusType `gorm:"column:status;not null;default:'Pending'"`
	StartedAt *string `gorm:"column:started_at"`
	EndedAt *string `gorm:"column:ended_at"`
	TimeLimitMinutes int `gorm:"column:time_limit_minutes;not null;default:50"`
	TournamentID uint `gorm:"column:tournament_id"`
}

func (m *TournamentRound) ToResponse() TournamentRoundResponse {
	return TournamentRoundResponse{
		ID:        m.ID,
		CreatedAt: m.CreatedAt,
		UpdatedAt: m.UpdatedAt,
		RoundNumber: m.RoundNumber,
		Status: m.Status,
		StartedAt: m.StartedAt,
		EndedAt: m.EndedAt,
		TimeLimitMinutes: m.TimeLimitMinutes,
		TournamentID: m.TournamentID,
	}
}

func (m *TournamentRound) ApplyUpdate(req TournamentRoundUpdateRequest) {
	if req.RoundNumber != nil { m.RoundNumber = *req.RoundNumber }
	if req.Status != nil { m.Status = *req.Status }
	if req.StartedAt != nil { m.StartedAt = req.StartedAt }
	if req.EndedAt != nil { m.EndedAt = req.EndedAt }
	if req.TimeLimitMinutes != nil { m.TimeLimitMinutes = *req.TimeLimitMinutes }
	if req.TournamentID != nil { m.TournamentID = *req.TournamentID }
}

func (m *TournamentRound) Start()  error {
	return fmt.Errorf("Start: not implemented")
}

func (m *TournamentRound) Complete()  error {
	return fmt.Errorf("Complete: not implemented")
}

func (m *TournamentRound) GeneratePairings()  error {
	return fmt.Errorf("GeneratePairings: not implemented")
}

func (m *TournamentRound) IsTimeExpired()  (bool, error) {
	return false, fmt.Errorf("IsTimeExpired: not implemented")
}

File: axum/src/handlers/tournaments/tournament_round.rs
use axum::{extract::{Path, Query, State}, http::StatusCode, Json};
use serde::Deserialize;
use sqlx::SqlitePool;
use crate::models::tournaments::tournament_round::{TournamentRound, TournamentRoundCreateRequest, TournamentRoundStatus};

type AppState = SqlitePool;

fn validate_tournament_round(payload: &TournamentRoundCreateRequest) -> Vec<String> {
    let mut errors: Vec<String> = Vec::new();
    if !((!(payload.ended_at.is_some()) || payload.ended_at.as_deref() > payload.started_at.as_deref())) { errors.push("Round end time must be after start time".to_string()); }
    if !((!(payload.status == TournamentRoundStatus::Completed) || payload.started_at.is_some())) { errors.push("Completed round must have a start time".to_string()); }
    if !(payload.round_number > 0) { errors.push("Round number must be greater than zero".to_string()); }
    if !(payload.time_limit_minutes > 0) { errors.push("Round time limit must be greater than zero".to_string()); }
    errors
}

#[derive(Deserialize)]
pub struct TournamentRoundListParams {
    pub skip: Option<i64>,
    pub limit: Option<i64>,
}

pub async fn list_tournament_round(
    State(pool): State<AppState>,
    Query(params): Query<TournamentRoundListParams>,
) -> Result<Json<Vec<TournamentRound>>, (StatusCode, String)> {
    let skip  = params.skip.unwrap_or(0);
    let limit = params.limit.unwrap_or(100).min(500);
    let rows = sqlx::query_as_unchecked!(TournamentRound, "SELECT * FROM tournament_rounds LIMIT $1 OFFSET $2", limit, skip)
        .fetch_all(&pool).await;
    rows.map(Json).map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))
}

pub async fn create_tournament_round(
    State(pool): State<AppState>,
    Json(payload): Json<TournamentRoundCreateRequest>,
) -> Result<(StatusCode, Json<TournamentRound>), (StatusCode, String)> {
    let errors = validate_tournament_round(&payload);
    if !errors.is_empty() { return Err((StatusCode::BAD_REQUEST, errors.join(", "))); }
    let row = sqlx::query_as_unchecked!(TournamentRound,
        "INSERT INTO tournament_rounds (round_number, status, started_at, ended_at, time_limit_minutes, tournament_id, created_at, updated_at) VALUES ($1, $2, $3, $4, $5, $6, datetime('now'), datetime('now')) RETURNING *",
        payload.round_number, payload.status, payload.started_at, payload.ended_at, payload.time_limit_minutes, payload.tournament_id
    ).fetch_one(&pool).await
    .map_err(|e| {
        if e.to_string().contains("UNIQUE") {
            (StatusCode::UNPROCESSABLE_ENTITY, "Value must be unique".to_string())
        } else {
            (StatusCode::INTERNAL_SERVER_ERROR, e.to_string())
        }
    })?;
    Ok((StatusCode::CREATED, Json(row)))
}

pub async fn get_tournament_round(
    State(pool): State<AppState>,
    Path(id): Path<i64>,
) -> Result<Json<TournamentRound>, (StatusCode, String)> {
    let row = sqlx::query_as_unchecked!(TournamentRound, "SELECT * FROM tournament_rounds WHERE id = $1", id)
        .fetch_optional(&pool).await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
        .ok_or((StatusCode::NOT_FOUND, "TournamentRound not found".to_string()))?;
    Ok(Json(row))
}

pub async fn start_tournament_round(
    State(pool): State<AppState>,
    Path(id): Path<i64>,
) -> Result<StatusCode, (StatusCode, String)> {
    let _row = sqlx::query_as_unchecked!(TournamentRound, "SELECT * FROM tournament_rounds WHERE id = $1", id)
        .fetch_optional(&pool).await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
        .ok_or((StatusCode::NOT_FOUND, "TournamentRound not found".to_string()))?;
    // TODO: implement start business logic
    Ok(StatusCode::OK)
}

pub async fn complete_tournament_round(
    State(pool): State<AppState>,
    Path(id): Path<i64>,
) -> Result<StatusCode, (StatusCode, String)> {
    let _row = sqlx::query_as_unchecked!(TournamentRound, "SELECT * FROM tournament_rounds WHERE id = $1", id)
        .fetch_optional(&pool).await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
        .ok_or((StatusCode::NOT_FOUND, "TournamentRound not found".to_string()))?;
    // TODO: implement complete business logic
    Ok(StatusCode::OK)
}

pub async fn generate_pairings_tournament_round(
    State(pool): State<AppState>,
    Path(id): Path<i64>,
) -> Result<StatusCode, (StatusCode, String)> {
    let _row = sqlx::query_as_unchecked!(TournamentRound, "SELECT * FROM tournament_rounds WHERE id = $1", id)
        .fetch_optional(&pool).await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
        .ok_or((StatusCode::NOT_FOUND, "TournamentRound not found".to_string()))?;
    // TODO: implement generate_pairings business logic
    Ok(StatusCode::OK)
}

pub async fn is_time_expired_tournament_round(
    State(pool): State<AppState>,
    Path(id): Path<i64>,
) -> Result<StatusCode, (StatusCode, String)> {
    let _row = sqlx::query_as_unchecked!(TournamentRound, "SELECT * FROM tournament_rounds WHERE id = $1", id)
        .fetch_optional(&pool).await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
        .ok_or((StatusCode::NOT_FOUND, "TournamentRound not found".to_string()))?;
    // TODO: implement is_time_expired business logic
    Ok(StatusCode::OK)
}

pub fn tournament_round_router() -> axum::Router<AppState> {
    axum::Router::new()
        .route("/api/tournament_rounds", axum::routing::get(list_tournament_round).post(create_tournament_round))
        .route("/api/tournament_rounds/:id", axum::routing::MethodRouter::new().get(get_tournament_round))
        .route("/api/tournament_rounds/:id/api/rounds/{id}/start", axum::routing::post(start_tournament_round))
        .route("/api/tournament_rounds/:id/api/rounds/{id}/complete", axum::routing::post(complete_tournament_round))
        .route("/api/tournament_rounds/:id/api/rounds/{id}/pairings", axum::routing::post(generate_pairings_tournament_round))
        .route("/api/tournament_rounds/:id/api/rounds/{id}/time-expired", axum::routing::get(is_time_expired_tournament_round))
}

#[cfg(test)]
mod tests {
    use super::*;
    use axum::body::Body;
    use axum::http::{Request, StatusCode};
    use serde_json::json;
    use tower::ServiceExt;
    use sqlx::SqlitePool;

    async fn setup_pool() -> SqlitePool {
        let pool = SqlitePool::connect(":memory:").await.unwrap();
        sqlx::migrate!("./migrations").run(&pool).await.unwrap();
        pool
    }

    fn app(pool: SqlitePool) -> axum::Router {
        tournament_round_router().with_state(pool)
    }

    #[tokio::test]
    async fn test_list_tournament_round() {
        let pool = setup_pool().await;
        let resp = app(pool).oneshot(
            Request::builder().uri("/api/tournament_rounds").body(Body::empty()).unwrap()
        ).await.unwrap();
        assert_eq!(resp.status(), StatusCode::OK);
    }

    #[tokio::test]
    async fn test_create_tournament_round() {
        let pool = setup_pool().await;
        let body = json!({
        "round_number": 2,
        "status": "Pending",
        "time_limit_minutes": 2,
        "tournament_id": 1
    });
        let resp = app(pool).oneshot(
            Request::builder()
                .method("POST")
                .uri("/api/tournament_rounds")
                .header("content-type", "application/json")
                .body(Body::from(body.to_string())).unwrap()
        ).await.unwrap();
        assert_eq!(resp.status(), StatusCode::CREATED);
    }

    #[tokio::test]
    async fn test_retrieve_tournament_round() {
        let pool = setup_pool().await;
        let resp = app(pool).oneshot(
            Request::builder().uri("/api/tournament_rounds/1").body(Body::empty()).unwrap()
        ).await.unwrap();
        assert!(resp.status() == StatusCode::NOT_FOUND || resp.status() == StatusCode::OK);
    }

}
